src/main/scheduler/scheduler.c

```diff
@@ -32,247 +33,301 @@
+        {
+            if(&cfTasks[ii] == currentTask)
+            {
+                taskId = ii;
+                break;
+            }
+        }
 
-    /* Calculate guard interval */
-    uint32_t maxNonRealtimeTaskTime = 0;
-    for (const cfTask_t *task = queueFirst(); task != NULL; task = queueNext()) {
-        if (task->staticPriority != TASK_PRIORITY_REALTIME) {
-            maxNonRealtimeTaskTime = MAX(maxNonRealtimeTaskTime, task->averageExecutionTime);
+        // Failed to find the task.
+        if(taskId < 0)
+        {
+            return;
         }
     }
 
-    realtimeGuardInterval = constrain(maxNonRealtimeTaskTime, REALTIME_GUARD_INTERVAL_MIN, REALTIME_GUARD_INTERVAL_MAX) + REALTIME_GUARD_INTERVAL_MARGIN;
-#if defined SCHEDULER_DEBUG
-    debug[2] = realtimeGuardInterval;
-#endif
-}
-
-#ifndef SKIP_TASK_STATISTICS
-void getTaskInfo(const int taskId, cfTaskInfo_t * taskInfo)
-{
+    // Fill the stats
     taskInfo->taskName = cfTasks[taskId].taskName;
-    taskInfo->isEnabled = queueContains(&cfTasks[taskId]);
+    taskInfo->isEnabled = cfTasks[taskId].isEnabled;
     taskInfo->desiredPeriod = cfTasks[taskId].desiredPeriod;
-    taskInfo->staticPriority = cfTasks[taskId].staticPriority;
+    taskInfo->priority = cfTasks[taskId].priority;
+    taskInfo->isWaitingToBeRan = cfTasks[taskId].isWaitingToBeRan;
+    taskInfo->lastExecutionTime = cfTasks[taskId].lastExecutionTime;
     taskInfo->maxExecutionTime = cfTasks[taskId].maxExecutionTime;
     taskInfo->totalExecutionTime = cfTasks[taskId].totalExecutionTime;
     taskInfo->averageExecutionTime = cfTasks[taskId].averageExecutionTime;
-    taskInfo->latestDeltaTime = cfTasks[taskId].taskLatestDeltaTime;
 }
-#endif
 
-void rescheduleTask(const int taskId, uint32_t newPeriodMicros)
+void updateTaskExecutionPeriod(const int taskId, uint32_t newPeriodMicros)
 {
-    if (taskId == TASK_SELF || taskId < (int)taskCount) {
+    if ((taskId == TASK_SELF && currentTask != NULL) || (taskId < (int)taskCount && taskId >= 0)) {
         cfTask_t *task = taskId == TASK_SELF ? currentTask : &cfTasks[taskId];
         task->desiredPeriod = MAX(100, newPeriodMicros);  // Limit delay to 100us (10 kHz) to prevent scheduler clogging
     }
 }
 
 void setTaskEnabled(const int taskId, bool enabled)
 {
-    if (taskId == TASK_SELF || taskId < (int)taskCount) {
+    if ((taskId == TASK_SELF && currentTask != NULL) || (taskId < (int)taskCount && taskId >= 0)) {
         cfTask_t *task = taskId == TASK_SELF ? currentTask : &cfTasks[taskId];
-        if (enabled && task->taskFunc) {
-            queueAdd(task);
-        } else {
-            queueRemove(task);
-        }
+        task->isEnabled = (enabled && task->taskFunc);
     }
 }
 
-uint32_t getTaskDeltaTime(const int taskId)
+uint32_t getTaskTimeSinceLastRun(const int taskId)
 {
-    if (taskId == TASK_SELF || taskId < (int)taskCount) {
+    if ((taskId == TASK_SELF && currentTask != NULL) || (taskId < (int)taskCount && taskId >= 0)) {
```

Comment:
i see this same 'if (..)' statement 3 times, perhaps it has a name and should be reused?